File: smoke_tests/kotlin/ssrf.kt
// Server-Side Request Forgery (SSRF) vulnerabilities in Kotlin
package com.example.security

import java.net.URL
import java.net.HttpURLConnection
import java.io.BufferedReader
import java.io.InputStreamReader
import okhttp3.OkHttpClient
import okhttp3.Request

class SsrfVulnerabilities {

    private val httpClient = OkHttpClient()

    // Test 1: Direct URL fetch
    fun fetchUrl(urlString: String): String {
        // VULNERABLE: User-controlled URL
        val url = URL(urlString)
        val connection = url.openConnection() as HttpURLConnection
        return connection.inputStream.bufferedReader().readText()
    }

    // Test 2: Image proxy
    fun proxyImage(imageUrl: String): ByteArray {
        // VULNERABLE: Fetching arbitrary URLs
        val url = URL(imageUrl)
        return url.readBytes()
    }

    // Test 3: Webhook URL
    fun sendWebhook(webhookUrl: String, payload: String): Int {
        // VULNERABLE: Webhook destination from user
        val url = URL(webhookUrl)
        val connection = url.openConnection() as HttpURLConnection
        connection.requestMethod = "POST"
        connection.doOutput = true
        connection.outputStream.write(payload.toByteArray())
        return connection.responseCode
    }

    // Test 4: URL from database
    fun fetchFromConfig(configKey: String): String {
        val targetUrl = getConfigValue(configKey)
        // VULNERABLE: URL from potentially tainted config
        val url = URL(targetUrl)
        return url.readText()
    }

    // Test 5: Partial URL construction
    fun fetchFromHost(hostname: String): String {
        // VULNERABLE: User controls hostname
        val url = URL("http://$hostname/api/data")
        return url.readText()
    }

    // Test 6: Port scanning
    fun checkPort(host: String, port: Int): Boolean {
        // VULNERABLE: Port from user
        return try {
            val url = URL("http://$host:$port/")
            val connection = url.openConnection() as HttpURLConnection
            connection.connectTimeout = 1000
            connection.connect()
            true
        } catch (e: Exception) {
            false
        }
    }

    // Test 7: OkHttp client
    fun fetchWithOkHttp(targetUrl: String): String? {
        // VULNERABLE: User URL with OkHttp
        val request = Request.Builder()
            .url(targetUrl)
            .build()
        return httpClient.newCall(request).execute().body?.string()
    }

    // Test 8: File protocol SSRF
    fun readResource(uri: String): String {
        // VULNERABLE: Can use file:// protocol
        val url = URL(uri)
        return url.readText()
    }

    // Test 9: URL from headers
    fun fetchFromHeader(request: HttpRequest): String {
        val targetUrl = request.getHeader("X-Target-URL")
        // VULNERABLE: URL from header
        return URL(targetUrl).readText()
    }

    // Test 10: Redirect following
    fun fetchWithRedirects(urlString: String): String {
        val url = URL(urlString)
        val connection = url.openConnection() as HttpURLConnection
        // VULNERABLE: Following redirects to internal
        connection.instanceFollowRedirects = true
        return connection.inputStream.bufferedReader().readText()
    }

    // Test 11: DNS rebinding
    fun fetchExternal(domain: String, path: String): String {
        // VULNERABLE: DNS can resolve to internal IP
        val url = URL("http://$domain/$path")
        return url.readText()
    }

    // Test 12: API endpoint construction
    fun callExternalApi(baseUrl: String, endpoint: String): String {
        // VULNERABLE: Both from user
        val url = URL("$baseUrl/$endpoint")
        return url.readText()
    }

    private fun getConfigValue(key: String): String = ""
}

interface HttpRequest {
    fun getHeader(name: String): String
}

File: smoke_tests/scala/ssrf.scala
// Server-Side Request Forgery (SSRF) vulnerabilities in Scala
package com.example.security

import java.net.{URL, HttpURLConnection}
import scala.io.Source

class SsrfVulnerabilities {

  // Test 1: Direct URL fetch
  def fetchUrl(urlString: String): String = {
    // VULNERABLE: User-controlled URL
    val url = new URL(urlString)
    Source.fromURL(url).mkString
  }

  // Test 2: Image proxy
  def proxyImage(imageUrl: String): Array[Byte] = {
    // VULNERABLE: Fetching arbitrary URLs
    val url = new URL(imageUrl)
    val connection = url.openConnection()
    val stream = connection.getInputStream
    Stream.continually(stream.read).takeWhile(_ != -1).map(_.toByte).toArray
  }

  // Test 3: Webhook URL
  def sendWebhook(webhookUrl: String, payload: String): Int = {
    // VULNERABLE: Webhook destination from user
    val url = new URL(webhookUrl)
    val connection = url.openConnection().asInstanceOf[HttpURLConnection]
    connection.setRequestMethod("POST")
    connection.setDoOutput(true)
    connection.getOutputStream.write(payload.getBytes)
    connection.getResponseCode
  }

  // Test 4: Partial URL construction
  def fetchFromHost(hostname: String): String = {
    // VULNERABLE: User controls hostname
    val url = new URL(s"http://$hostname/api/data")
    Source.fromURL(url).mkString
  }

  // Test 5: Port scanning
  def checkPort(host: String, port: Int): Boolean = {
    // VULNERABLE: Port from user
    try {
      val url = new URL(s"http://$host:$port/")
      val connection = url.openConnection().asInstanceOf[HttpURLConnection]
      connection.setConnectTimeout(1000)
      connection.connect()
      true
    } catch {
      case _: Exception => false
    }
  }

  // Test 6: Akka HTTP client (conceptual)
  def akkaHttpRequest(targetUrl: String): String = {
    // VULNERABLE: User URL with Akka HTTP
    s"Http().singleRequest(HttpRequest(uri = $targetUrl))"
  }

  // Test 7: File protocol SSRF
  def readResource(uri: String): String = {
    // VULNERABLE: Can use file:// protocol
    val url = new URL(uri)
    Source.fromURL(url).mkString
  }

  // Test 8: Redirect following
  def fetchWithRedirects(urlString: String): String = {
    val url = new URL(urlString)
    val connection = url.openConnection().asInstanceOf[HttpURLConnection]
    // VULNERABLE: Following redirects to internal
    connection.setInstanceFollowRedirects(true)
    Source.fromInputStream(connection.getInputStream).mkString
  }

  // Test 9: DNS rebinding
  def fetchExternal(domain: String, path: String): String = {
    // VULNERABLE: DNS can resolve to internal IP
    val url = new URL(s"http://$domain/$path")
    Source.fromURL(url).mkString
  }

  // Test 10: sttp client
  def sttpRequest(targetUrl: String): String = {
    // VULNERABLE: User URL with sttp
    s"basicRequest.get(uri\"$targetUrl\")"
  }

  // Test 11: API endpoint construction
  def callExternalApi(baseUrl: String, endpoint: String): String = {
    // VULNERABLE: Both from user
    val url = new URL(s"$baseUrl/$endpoint")
    Source.fromURL(url).mkString
  }

  // Test 12: Play WS client
  def playWsRequest(targetUrl: String): String = {
    // VULNERABLE: User URL with Play WS
    s"ws.url($targetUrl).get()"
  }
}
